doc/environment.adoc

```diff
@@ -0,0 +1,247 @@
+
+The `seen` table is a lookup for already copied tables, which are used as keys. The value is the copy. By providing `[table] = table` entries for certain tables, you can prevent them from being copied.
+
+==== `table.insert_all(t, other)`
+
+Adds all the list entries of `other` to `t` (list part concatenation).
+
+==== `table.key_value_swap(t)`
+
+Returns a new table with the keys of `t` as values and the corresponding values as keys. If a value occurs multiple times in `t`, any of the keys might be the value in the resulting table.
+
+==== `table.shuffle(t, from, to, random)`
+
+Performs a Fisher-Yates shuffling on the specified range of the list part of `t`.
+
+* `from`: Inclusive starting index of the range to be shuffled. Defaults to the first item of the list part if falsy.
+* `to`: Inclusive end index of the range to be shuffled. Defaults to the last item of the list part if falsy.
+* `random`: A `function(from, to)` that returns a random integer in the specified range, with both `from` and `to` inclusive. Defaults to `math.random` if falsy.
+
+Returns nothing.
+
+== LuaJIT extensions
+
+Minetest builds compiled with LuaJIT (`ENABLE_LUAJIT=1`) provide the https://luajit.org/extensions.html[LuaJIT extensions]. These include syntactical Lua 5.2 language features like `goto`, which will lead to a syntax error on PUC Lua 5.1. Hex escapes will be converted into the raw characters by PUC Lua 5.1.
+
+== Common extensions
+
+https://bitop.luajit.org/[LuaJIT's `bit` library] is made available for both PUC Lua and LuaJIT builds. It must not be required, as this will lead to a crash in a secure environment as documented below; in an insecure environment, it is simply unneeded. 
+
+== Secure environment whitelists
+
+In the secure environment, the following builtin Lua(JIT) libraries and library functions are whitelisted:
+
+* `_VERSION`
+* Garbage collection: `collectgarbage`
+* Cooperative multithreading: `coroutine`
+* Error handling:
+** `assert`
+** `error`
+** `pcall`
+** `xpcall`
+* Function environments:
+** `setfenv`
+** `getfenv`
+* `math`
+* `string`
+* Tables:
+** `table`
+** Iteration:
+*** `next`
+*** `pairs`
+*** `ipairs`
+* Metatables:
+** `setmetatable`
+** Not available to CSM: `getmetatable`
+** Raw methods:
+*** `rawset`
+*** `rawget`
+*** `rawequals`
+* Varargs:
+** `select`
+** `unpack`
+* Conversion:
+** `tostring`
+** `tonumber`
+* `type`
+* Output: `print`
+
+Some library tables are restricted by whitelists as well:
+
+* Not available to CSM: `io`
+** `read`
+** `write`
+** `flush`
+** `close`
+** `type`
+* `os`: Mostly time-related functions
+** `clock`
+** `date`
+** `difftime`
+** `time`
+** Not available to CSM:
```

Comment:
I feel like there's a better way to make "Not available to CSM" a note here rather than the first information the reader sees, but I'm not sure what.